Homework

sdziuda opened the pull request:
Szymon Dziuda

sdziuda committed:
Added implementation

Added implementation with a couple of clients and random generated action history

sdziuda committed:
Improvements

Changed from this.clients to this.getAllClients()

avan1235 commented on src/main/java/pl/edu/mimuw/Bank.java:
```diff
@@ -1,20 +1,84 @@
+    this.clients = new ArrayList<BankClient>();
+  }
+
+  /**
```
Super 👌🏻

avan1235 commented on src/main/java/pl/edu/mimuw/BankClient.java:
```diff
@@ -1,35 +1,143 @@
 
   public List<BankAction> getActionHistory() {
-    throw new IllegalStateException("TODO");
+    return this.actionHistory;
```
Może by jednak zwracać kopie tej listy, bo jest ona modyfikowalna i wystawiamy w ten sposób prywatne pole do dowolnej modyfikacji?

avan1235 commented on src/main/java/pl/edu/mimuw/BankClient.java:
```diff
@@ -1,35 +1,143 @@
+  /**
+   * This method adds a couple of random actions to given client's action history.
+   */
+  public void addRandomActions() {
```
To powinna być raczej metoda statyczna w Main 😉 Klasy niech zawierają tylko metody dotyczącej ich dziedziny

avan1235 commented on src/main/java/pl/edu/mimuw/BankClient.java:
```diff
@@ -1,35 +1,143 @@
+    double result = 0;
+
+    for (var action : this.actionHistory) {
+      if (action instanceof DepositBankAction) {
```
Chcemy raczej unikać ile się da używania instanceof w programowaniu obiektowym .
Moim zdaniem warto żebyś przemyślał/spróbował może zmienić interfejs, tak żeby nie trzeba tu było sprawdzać typu obiektu - może akcje powinny mieć inne metody, tak żebyś mógł na każdej coś zawołać i odddzilenie dla wszystkich policzyć włożone i zyskane pieniądzę? Przemyśl proszę 😉

avan1235 commented on src/main/java/pl/edu/mimuw/BankClient.java:
```diff
@@ -1,35 +1,143 @@
+
+    final var actionsList = new ArrayList<BankAction>(this.actionHistory);
+    actionsList.stream()
+      .sorted(Comparator.comparing(BankAction::getStartTimestamp))
```
Mega pattern 💪🏻

avan1235 commented on src/main/java/pl/edu/mimuw/Bank.java:
```diff
@@ -1,20 +1,84 @@
 
 public class Bank {
 
+  private final List<BankClient> clients;
```
może ten bank to by miał jakąś nazwę chociaż ;)

avan1235 commented on src/main/java/pl/edu/mimuw/Bank.java:
```diff
@@ -1,20 +1,84 @@
+    final var moneyLoaned = new BigDecimal(String.valueOf(this.getMoneyLoaned()))
+      .setScale(2, RoundingMode.HALF_UP);
+
+    sb.append("Number of clients in the bank: ").append(this.getNumberOfClients()).append("\n");
```
proponuje
```
sb.append("Number of clients in the bank: ").append(this.getNumberOfClients()).append("\n")
  .append("Money deposited in the bank: ").append(moneyDeposited).append("\n")
  .append("Money loaned from the bank: ").append(moneyLoaned).append("\n\n");
```

avan1235 commented on src/main/java/pl/edu/mimuw/Bank.java:
```diff
@@ -1,20 +1,84 @@
+    final var clientsList = this.getAllClients();
+    clientsList.stream()
+      .sorted(Comparator.comparing(BankClient::getAccountNumber))
+      .forEach(sb::append);
```
super! a nawet można `this.getAllClients().stream()...`

avan1235 commented on src/main/java/pl/edu/mimuw/Bank.java:
```diff
@@ -1,20 +1,84 @@
-    throw new IllegalStateException("TODO: remember not to return internal state list as it can be modified");
+    final var clients = new ArrayList<BankClient>();
+
+    for (var client : this.clients) {
```
np mozna 
```
this.clients.stream().map(BankClient::new).forEach(clients::add)
```
jak juz szalejesz ze streamami

avan1235 commented on src/main/java/pl/edu/mimuw/Bank.java:
```diff
@@ -1,20 +1,84 @@
+  /**
+   * @return sum of all money loaned from the bank.
+   */
+  public double getMoneyLoaned() {
```
to jest raczej coś co powinno się dziać przy dodawaniu klientów - mieć pola z tą sumą żeby za każdym razem nie sumować

avan1235 commented on src/main/java/pl/edu/mimuw/BankClient.java:
```diff
@@ -1,35 +1,143 @@
+    this.surname = surname;
+    this.age = age;
+    this.accountNumber = accountNumber;
+    this.actionHistory = new ArrayList<>(actionHistory);
```
racja, dobrze zauważyłeś że należy kopiować, nie wiem czy o tym mówiełem  👍🏻

avan1235 commented on src/main/java/pl/edu/mimuw/Main.java:
```diff
@@ -1,18 +1,38 @@
+
+    final var clientsList = Arrays.asList(clients);
+
+    clientsList.stream().forEach(bank::addClient);
```
`Arrays.stream(clients)...`

avan1235 reviewed (commented):
podoba mi się, zobacz na drobne uwagi/sugestie. 1 pkt dla Ciebie

sdziuda committed:
Further improvements

Added some improvements as suggested in the comments to previous commits

sdziuda committed:
Changed visibility in BankAction

Changed constructor's and method's visibility from public to protected (as it was unnecessary)
